src/js/d3_svg_cell_renderer.js

```diff
@@ -0,0 +1,207 @@
+var utils = require('./utils');
+var $ = require('jquery');
+var _ = require('underscore');
+
+// TODO: handle accessing config properties cleaner
+
+function D3SVGRuleSet(renderer) {
+	var self = this;
+	self.rule_map = {};
+	self.renderer = renderer;
+
+	self.addRule = function(condition, d3_shape, attrs, z_index) {
+		var rule_id = Object.keys(self.rule_map).length;
+		if (z_index === undefined) {
+			z_index = rule_id;
+		}
+		self.rule_map[rule_id] = new D3SVGRule(rule_id, self.renderer, condition, d3_shape, attrs, z_index);
+		return rule_id;
+	};
+
+	self.removeRule = function(rule_id) {
+		delete self.rule_map[rule_id];
+	}
+
+	self.getRules = function() {
+		// returns a list of lists of rules to render in the given order
+		var z_map = {};
+		_.each(self.rule_map, function(rule, rule_id) {
+			z_map[rule.z_index] = z_map[rule.z_index] || [];
+			z_map[rule.z_index].push(rule);
+		});
+		return _.map(Object.keys(z_map).sort(), 
+			function(z) { return z_map[z];});
+	};
+
+	self.fromJSON = function(json_rules) {
+		self.rule_map = {};
+		_.each(json_rules, function(rule) {
+			self.addRule(rule.condition, rule.d3_shape, rule.attrs, rule.z_index);
+		});
+	};
+}
+
+function D3SVGRule(rule_id, renderer, condition, d3_shape, attrs, z_index) {
```

Comment:
1) It is self-documenting - instead of passing around a nude map and saying "this is a rule," you can point directly at the code and know what's going on when you say new D3SVGRule().

2) Because a higher order function can only consist of one function. You can see right here there is more than one function associated with the rule: there is ```apply``` and there is ```filterData```